code/go/0chain.net/smartcontract/storagesc/allocation.go

```diff
@@ -1235,38 +1235,43 @@ func (sc *StorageSmartContract) canceledPassRates(alloc *StorageAllocation,
 	passRates = make([]float64, 0, len(alloc.BlobberDetails))
 	var failed, successful int64 = 0, 0
 
-	allocChallenge, err := sc.getAllocationChallenge(alloc.ID, balances)
-	switch err {
-	case util.ErrValueNotPresent:
-	case nil:
-		for _, c := range allocChallenge.Challenges {
-			blobberID := c.BlobberID
-			d, ok := alloc.BlobberMap[blobberID]
-			if !ok {
-				continue
-			}
+	for _, d := range alloc.BlobberDetails {
 
-			if d.Stats == nil {
-				d.Stats = new(StorageAllocationStats) // make sure
+		var foundChallenge bool
+		bc, err := sc.getBlobberChallenge(d.BlobberID, balances)
+		switch err {
+		case nil:
+			for i := 0; i < len(bc.Challenges); i++ {
+				chall := bc.Challenges[i]
+				if chall.AllocationID == alloc.ID {
```

Comment:
We only need to check challenges that has expired, since all challenges are already sorted by timestamp in asc, so we can break once detected a challenge that is not expired. The persudo code could be:

```go
for i := 0; i < len(challenges); i++ {
    clg := challenges[i]
    if !isExpired(clg, now) {
        break
    }
    
    if clg.AllocationID != alloc.ID {
        continue
    }

    // found the allocation 
    //  remove the challenge and update stats...
}
```